libkbfs/bserver_tlf_storage.go

```diff
@@ -0,0 +1,409 @@
+	}
+
+	keyServerHalfPath := s.buildKeyServerHalfPath(id)
+	buf, err := ioutil.ReadFile(keyServerHalfPath)
+	if os.IsNotExist(err) {
+		return nil, BlockCryptKeyServerHalf{},
+			BServerErrorBlockNonExistent{}
+	} else if err != nil {
+		return nil, BlockCryptKeyServerHalf{}, err
+	}
+
+	var serverHalfData [32]byte
+	if len(buf) != len(serverHalfData) {
+		return nil, BlockCryptKeyServerHalf{},
+			fmt.Errorf("Corrupt server half data file %s",
+				keyServerHalfPath)
+	}
+	copy(serverHalfData[:], buf)
+	serverHalf := MakeBlockCryptKeyServerHalf(serverHalfData)
+	return data, serverHalf, nil
+}
+
+func (s *bserverTlfStorage) getRefEntriesLocked(id BlockID) (
+	map[BlockRefNonce]blockRefEntry, error) {
+	refsPath := s.buildRefsPath(id)
+	// Let caller handle os.IsNotExist(err) case.
+	refInfos, err := ioutil.ReadDir(refsPath)
+	if err != nil {
+		return nil, err
+	}
+
+	refs := make(map[BlockRefNonce]blockRefEntry)
+	for _, refInfo := range refInfos {
+		var refNonce BlockRefNonce
+		buf, err := hex.DecodeString(refInfo.Name())
+		if err != nil {
+			return nil, err
+		}
+		if len(buf) != len(refNonce) {
+			return nil, fmt.Errorf(
+				"Invalid ref nonce file %s", refInfo.Name())
+		}
+		copy(refNonce[:], buf)
+
+		refEntry, err := s.getRefEntryLocked(id, refNonce)
+		if err != nil {
+			return nil, err
+		}
+		refs[refNonce] = refEntry
+	}
+
+	return refs, nil
+}
+
+func (s *bserverTlfStorage) getAll() (
+	map[BlockID]map[BlockRefNonce]blockRefLocalStatus, error) {
+	res := make(map[BlockID]map[BlockRefNonce]blockRefLocalStatus)
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	if s.isShutdown {
+		return nil, errBserverTlfStorageShutdown
+	}
+
+	levelOneInfos, err := ioutil.ReadDir(s.dir)
+	if err != nil {
+		return nil, err
+	}
+
+	// Ignore non-dirs while traversing below.
+
+	// TODO: Tighten up checks below if we ever use this for
+	// anything other than tests.
+
+	for _, levelOneInfo := range levelOneInfos {
+		if !levelOneInfo.IsDir() {
+			continue
+		}
+
+		levelOneDir := filepath.Join(s.dir, levelOneInfo.Name())
+		levelTwoInfos, err := ioutil.ReadDir(levelOneDir)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, levelTwoInfo := range levelTwoInfos {
+			if !levelTwoInfo.IsDir() {
+				continue
+			}
+
+			idStr := levelOneInfo.Name() + levelTwoInfo.Name()
+			id, err := BlockIDFromString(idStr)
+			if err != nil {
+				return nil, err
+			}
+
+			_, ok := res[id]
+			if ok {
+				return nil, fmt.Errorf(
+					"Multiple dir entries for block %s", id)
+			}
+
+			res[id] = make(map[BlockRefNonce]blockRefLocalStatus)
+
+			refs, err := s.getRefEntriesLocked(id)
+			if err != nil {
+				return nil, err
+			}
+
+			for ref, refEntry := range refs {
+				res[id][ref] = refEntry.Status
+			}
+		}
+	}
+
+	return res, nil
+}
+
+func (s *bserverTlfStorage) putRefEntryLocked(
+	id BlockID, refEntry blockRefEntry) error {
+	buf, err := s.codec.Encode(refEntry)
+	if err != nil {
+		return err
+	}
+	// TODO: Add integrity-checking?
```

Comment:
Yeah, it's weird that the memory server checks but this one doesn't.  Can probably factor some of it out into a common helper.